northstar/src/runtime/fork/init/mod.rs

```diff
@@ -130,26 +235,42 @@ impl Init {
                         util::set_parent_death_signal(Signal::SIGKILL);
 
-                        // /dev/null/ is opened cloexec
-                        let dev_null = nix::fcntl::open(
-                            "/dev/null",
-                            nix::fcntl::OFlag::O_RDONLY | nix::fcntl::OFlag::O_CLOEXEC,
-                            nix::sys::stat::Mode::empty(),
-                        )
-                        .expect("Failed to open /dev/null");
-                        unistd::dup2(dev_null, nix::libc::STDIN_FILENO).expect("Failed to dup2");
-                        unistd::dup2(stdout, nix::libc::STDOUT_FILENO).expect("Failed to dup2");
-                        unistd::dup2(stderr, nix::libc::STDERR_FILENO).expect("Failed to dup2");
-
-                        unistd::close(stdout).expect("Failed to close stdout after dup2");
-                        unistd::close(stderr).expect("Failed to close stderr after dup2");
+                        match out {
+                            Out::Pipe { stdout, stderr } => {
+                                // /dev/null/ is opened cloexec
+                                let dev_null = nix::fcntl::open(
+                                    "/dev/null",
+                                    nix::fcntl::OFlag::O_RDONLY | nix::fcntl::OFlag::O_CLOEXEC,
+                                    nix::sys::stat::Mode::empty(),
+                                )
+                                .expect("Failed to open /dev/null");
+                                unistd::dup2(dev_null, nix::libc::STDIN_FILENO)
+                                    .expect("Failed to dup2");
+                                unistd::dup2(stdout, nix::libc::STDOUT_FILENO)
+                                    .expect("Failed to dup2");
+                                unistd::dup2(stderr, nix::libc::STDERR_FILENO)
+                                    .expect("Failed to dup2");
+                                unistd::close(stdout).expect("Failed to close");
+                                unistd::close(stderr).expect("Failed to close");
+                            }
+                            Out::Pty { slave } => {
+                                unistd::setsid().unwrap();
+                                if unsafe { nix::libc::ioctl(slave, nix::libc::TIOCSCTTY) } < 0 {
+                                    panic!("Failed to TIOCSCTTY");
+                                }
+                                unistd::dup2(slave, nix::libc::STDIN_FILENO)
+                                    .expect("Failed to dup2");
+                                unistd::dup2(slave, nix::libc::STDOUT_FILENO)
+                                    .expect("Failed to dup2");
+                                unistd::dup2(slave, nix::libc::STDERR_FILENO)
+                                    .expect("Failed to dup2");
+                                unistd::close(slave).expect("Failed to close");
+                            }
```

Comment:
What's the reason to distinguish the io and pty cases? What would be bad with using *always* a pty?

Reply:
There is a subtle detail between the two. To set the pty as the controlling tty of the process, it has to be placed in a separated session with `setsid()`. In that case, our `kill` command will fail to kill the process with the `-pid` trick because that only works for processes in the same session :weary: